### PR TITLE
Fix slice

### PR DESCRIPTION
**Pull request by simonJJJ**
There are two problems of slice op. The first is the different behaviour in eager mode with PyTorch:
b2033b45d76272326dd709286237a0d
after fix:
6c038611543bfcd0eeffe6c64ca20b7
The second is that lazy don't support scalar, So I add a lazy guard to fix it.

### CODE REVIEW

File: oneflow/user/ops/slice_op.cpp [hunk 1]
```diff
@@ -38,10 +39,17 @@ Maybe<void> InferSliceOpTensorDesc(user_op::InferContext* ctx) {
   CHECK_EQ_OR_RETURN(step_vec.size(), ndim);
 
   // slice a 1-dim tensor will return a 0-dim or 1-dim tensor
-  if (x_shape.NumAxes() == 1) {
-    const int64_t start = start_vec.at(0);
-    const int64_t stop = stop_vec.at(0);
+  if (x_shape.NumAxes() == 1 && !LazyMode::is_enabled()) {
```
Comment (strint) [inline oneflow/user/ops/slice_op.cpp:42]: 这一段是专门为 1-D tensor + eager写的？ lazy的 1-D不处理？

File: oneflow/core/thread/thread.cpp [hunk 2]
```diff
@@ -49,6 +50,7 @@ void Thread::AddTask(const TaskProto& task) {
 }
 
 void Thread::PollMsgChannel() {
+  auto lazy_mode_enabled_guard = LazyMode::Guard(/* is_enabled */ true);
```
Comment (strint) [inline oneflow/core/thread/thread.cpp:53]: 这个处理感觉需要讨论下。目前看单纯是为了slice的0-D tensor而写。

File: oneflow/user/ops/slice_op.cpp (hunk 1 above)
Comment (simonJJJ) [reply oneflow/user/ops/slice_op.cpp:42]: 是的

File: oneflow/core/thread/thread.cpp (hunk 2 above)
Comment (simonJJJ) [reply oneflow/core/thread/thread.cpp:53]: 是的

File: oneflow/api/python/framework/tensor.cpp [hunk 3]
```diff
@@ -263,7 +263,7 @@ Maybe<Tensor> NewTensor(py::args args, py::kwargs kwargs, Symbol<DType> desired_
                         bool treat_single_int_as_size) {
   // NOTE(chengcheng): flow.Tensor or flow.tensor ONLY created by EagerTensor now.
   //  even if in nn.Graph build (module forward function), if you create a flow.Tensor,
-  //  its a eager tensor by Run functional::Empty() in LazyMode::Grad(false)
+  //  its a eager tensor by Run functional::Empty() in LazyMode::Gurad(false)
```
Comment (chengtbf) [inline oneflow/api/python/framework/tensor.cpp:266]: Guard 拼错了

File: oneflow/user/ops/slice_op.cpp [hunk 4]
```diff
@@ -38,10 +39,17 @@ Maybe<void> InferSliceOpTensorDesc(user_op::InferContext* ctx) {
...
+    const int64_t dim_size = x_shape.At(0);
+    int64_t start = start_vec.at(0);
+    int64_t stop = stop_vec.at(0);
     const int64_t step = step_vec.at(0);
+    if (dim_size == 0 || start == stop) {
+      *ctx->OutputShape("y", 0) = Shape({0});
```
Comment (chengtbf) [inline oneflow/user/ops/slice_op.cpp:48]: 理论上 这里应该 `Shape({})` ，Scalar Tensor 的 shape 是 空，不是 num axes = 1 吧。 

这里之所以写成  `Shape({0})`  是为了让别的地方可以绕过去？

Comment (simonJJJ) [reply oneflow/user/ops/slice_op.cpp:48]: 这里是为了对齐pytorch, pytorch的shape:
```import torch
a = torch.randn(10)
b = a[3:3]
b.shape
# torch.Size([0])

File: oneflow/core/thread/thread.cpp [hunk 5]
```diff
@@ -16,6 +16,7 @@ limitations under the License.
 #include "oneflow/core/thread/thread.h"
 #include "oneflow/core/job/runtime_context.h"
 #include "oneflow/core/job/global_for.h"
+#include "oneflow/core/job/lazy_mode.h"
```
Comment (strint) [inline oneflow/core/thread/thread.cpp:19]: 这个也去掉？

File: oneflow/user/ops/slice_op.cpp [hunk 6]
```diff
@@ -14,6 +14,7 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 #include "oneflow/core/framework/framework.h"
+#include "oneflow/core/job/lazy_mode.h"
```
Comment (strint) [inline oneflow/user/ops/slice_op.cpp:17]: 这个也去掉？

File: oneflow/user/ops/slice_op.cpp [hunk 7]
```diff
@@ -39,20 +40,24 @@ Maybe<void> InferSliceOpTensorDesc(user_op::InferContext* ctx) {
...
+    start = RegulateSliceStart(start, dim_size);
+    stop = RegulateSliceStop(stop, dim_size);
     const int64_t diff = (step > 0) ? (stop - start - 1) : (stop - start + 1);
     const int64_t len = diff / step + 1;
     CHECK_GE_OR_RETURN(len, 1);
     if (len == 1) {
       // return a 0-dim tensor
-      DimVector zero_dim_vec(0);
-      *ctx->OutputShape("y", 0) = Shape(zero_dim_vec);
+      *ctx->OutputShape("y", 0) = Shape({});
```
Comment (strint) [inline oneflow/user/ops/slice_op.cpp:58]: slice是无法产生0-dim tensor的吧。

只有index可以产生？

```
>>> t = torch.rand(10)
>>> t
tensor([0.7466, 0.9933, 0.3019, 0.2546, 0.7142, 0.5676, 0.5144, 0.7639, 0.9051,
        0.4749])
>>> s = t[1:2]
>>> s
tensor([0.9933])
>>> s.shape
torch.Size([1])
>>> s1 = t[1]
>>> s1
tensor(0.9933)
>>> s1.shape
torch.Size([])
```

File: python/oneflow/test/graph/test_graph_scalar.py [hunk 8]
```diff
@@ -151,6 +151,84 @@ def build(self, x):
...
+
+    class MyModule(flow.nn.Module):
+        def __init__(self):
+            super().__init__()
+            self.weight = flow.nn.Parameter(
+                flow.tensor([1.0, 2.0, 3.0, 4.0], device=device)
+            )
+
+        def forward(self, x):
+            return x * self.weight[3]
```
Comment (strint) [inline python/oneflow/test/graph/test_graph_scalar.py:165]: 这里也会调用到slice么

File: python/oneflow/test/graph/test_graph_scalar.py [hunk 9]
```diff
@@ -151,6 +151,84 @@ def build(self, x):
...
+        x = flow.tensor(i * 1.0, device=device, requires_grad=False)
+        of_eager_out = my_module(x)
+        of_eager_out.backward()
+        of_sgd.step()
+        of_sgd.zero_grad()
+        eager_out_list.append(of_eager_out)
+
+    lazy_module = MyModule()
+
+    class ScalarTrainGraph(flow.nn.Graph):
```
Comment (strint) [inline python/oneflow/test/graph/test_graph_scalar.py:207]: 这些加到test_slice.py里面是不是比较合适

File: oneflow/user/ops/slice_op.cpp (hunk 7 above)
Comment (simonJJJ) [reply oneflow/user/ops/slice_op.cpp:58]: 是因为index本质调用了slice

File: python/oneflow/test/graph/test_graph_scalar.py (hunk 9 above)
Comment (simonJJJ) [reply python/oneflow/test/graph/test_graph_scalar.py:207]: 都可以，我没有意见

File: python/oneflow/test/graph/test_graph_scalar.py (hunk 8 above)
Comment (simonJJJ) [reply python/oneflow/test/graph/test_graph_scalar.py:165]: index内部会调用slice

File: python/oneflow/test/modules/test_0_dim_tensor.py [hunk 10]
```diff
@@ -86,6 +164,8 @@ def test_0_dim_tensor(test_case):
             _test_scalar_mul,
             _test_slice,
             _test_slice_backward,
+            _test_slice_scalar_graph,
+            _test_slice_scalar_train_graph,
```
Comment (strint) [inline python/oneflow/test/modules/test_0_dim_tensor.py:168]: 这里没有测试  tensor[i:i]这种情况？

**Review by strint — APPROVED** [review]
